Add compress zlib

masahide opened the pull request.

masahide committed:
boltdb をやめてjsonで保存

masahide committed:
直した

masahide committed:
test修正

masahide committed:
削除

masahide committed:
db作成後ポジションを保存する

masahide committed:
add json struct values

masahide committed:
追加忘れのファイルを実装した

masahide committed:
ReadAllでサイズを返すように

masahide committed:
path 設定忘れ

masahide committed:
typo

masahide committed:
zlib writer

masahide committed:
BestCompressionにした

masahide committed:
途中

masahide committed:
bin encoder,decoder

masahide committed:
catdbを修正

masahide committed:
Merge branch 'add-compress-zlib' of github.com:masahide/ftailer into add-compress-zlib

masahide committed:
修正

masahide committed:
関数名をリファクタ

masahide committed:
不要変数削除

masahide committed:
同時実行制御引数を追加

ando-masaki commented on core/db.go:
```diff
@@ -129,29 +132,279 @@ func (db *DB) Delete(ext string) error {
+		line++
+		if db.bin {
+			row, err = decodeRow(db.file)
+			row.Pos.Name = db.Pos.Name
```
row := &Row{}
としているだけなので
いきなり
row.Pos.Name
とするとパニックが起こる。

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x18 pc=0x4fd450]

goroutine 24 [running]:
github.com/masahide/ftailer/core.(*FtailDB).ReadAll(0xc820346a00, 0x2b84ff2f5248, 0xc82000a380, 0x2b84ff2f5270, 0xc8200be0d8, 0x0, 0x0)
      /home/ando_masaki/go/src/github.com/masahide/ftailer/core/db.go:262 +0x200
...
```

ando-masaki commented on tailex/tailex.go:
```diff
@@ -109,6 +111,8 @@ func (c *TailEx) tailFileSyncLoop(ctx context.Context) {
 // Glob検索で見つかるまで 1*time.Secondでpolling
 func (c *TailEx) GlobSearchLoop(ctx context.Context, pathFmt string) (string, error) {
+	c.WorkLimit <- true
+	defer func() { <-c.WorkLimit }()
```
tailすべきファイルが存在しない場合に、
処理が進まなくなったが、
ここが原因？

```
No such file.:GlobSearch s:'/var/log/iptables/iptables.log', /var/log/iptables/iptables.log
```

masahide committed:
2点修正

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x18 pc=0x4fd450]

goroutine 24 [running]:
github.com/masahide/ftailer/core.(*FtailDB).ReadAll(0xc820346a00,
0x2b84ff2f5248, 0xc82000a380, 0x2b84ff2f5270, 0xc8200be0d8, 0x0, 0x0)
      /home/ando_masaki/go/src/github.com/masahide/ftailer/core/db.go:262
      +0x200
      ...
```

tailすべきファイルが存在しない場合に、処理が進まない。

masahide committed:
エラー判定を追加

masahide committed:
エラー処理を追加

masahide committed:
EOF判定ミスを修正

ando-masaki commented on core/db.go:
```diff
@@ -129,29 +132,290 @@ func (db *DB) Delete(ext string) error {
+	if err != nil {
+		return nil, err
+	}
+	pos.Offset = p.Offset
```
```
@40000000565420261f25983c core.Position{                                                                                                                                                                                            [190/9197]
@40000000565420261f25983c   Name:       "/product/monsters/server/codeigniter/application/logs/10.9.10.26_20151124.json_result_error.log",
@40000000565420261f25a3f4   CreateAt:   2015-11-24 05:22:58 Local,
@40000000565420261f25a7dc   Offset:     17346,
@40000000565420261f25a7dc   HeadHash:   "",
@40000000565420261f25abc4   HashLength: 0,
@40000000565420261f25b394 } (*core.Position)(nil)
@40000000565420261f3ffe0c panic: runtime error: invalid memory address or nil pointer dereference
@40000000565420261f4001f4 [signal 0xb code=0x1 addr=0x28 pc=0x4fd4ab]
@40000000565420261f401d4c 
@40000000565420261f402134 goroutine 32 [running]:
@40000000565420261f4038a4 github.com/masahide/ftailer/core.(*FtailDB).lastPostion(0xc820358e40, 0xc8204fed80, 0x5f, 0xecde56ca2, 0x0, 0xdc8720, 0x43c2, 0x0, 0x0, 0x0, ...)
@40000000565420261f40b98c       /home/ando_masaki/go/src/github.com/masahide/ftailer/core/db.go:222 +0x1fb
```

masahide committed:
ヘッダだけのファイルの場合readallはヘッダ情報を採用するように

masahide committed:
修正

masahide committed:
修正

masahide committed:
出力されない問題を修正
